resources/migrations/001_update_migrations.yaml

```diff
@@ -6494,6 +6483,238 @@ databaseChangeLog:
+            sql: >-
+              ALTER TABLE metabase_field MODIFY
+              name VARCHAR(254)
+              CHARACTER SET utf8mb4
+              COLLATE utf8mb4_unicode_ci;
+
+  # metabase_field.is_defective_duplicate is a new column that indicates whether a field is a defective duplicate field
+  # that should never have been created under Postgres' `idx_uniq_field_table_id_parent_id_name_2col` constraint, but
+  # was allowed to be created in MySQL or H2.
+  - changeSet:
+      id: v49.2024-06-27T00:00:01
+      author: calherries
+      comment: Add metabase_field.is_defective_duplicate
+      changes:
+        - addColumn:
+            tableName: metabase_field
+            columns:
+              - column:
+                  name: is_defective_duplicate
+                  type: ${boolean.type}
+                  remarks: "Indicates whether column is a defective duplicate field that should never have been created."
+                  constraints:
+                    nullable: false
+                  defaultValue: false
+
+  - changeSet:
+      id: v49.2024-06-27T00:00:02
+      author: calherries
+      comment: Populate metabase_field.is_defective_duplicate
+      changes:
+        - sql:
+            # idx_uniq_field_table_id_parent_id_name_2col would prevent defective duplicates with Postgres but it
+            # can be removed by upgrading from 49 and downgrading again. We need to populate is_defective_duplicate
+            # in that case.
+            dbms: postgresql
+            sql: >-
+              WITH ranked_fields AS (
+                  SELECT
+                      mf.id,
+                      ROW_NUMBER() OVER (
+                          PARTITION BY mf.table_id, mf.name, mf.parent_id
+                          ORDER BY
+                              CASE WHEN mf.active THEN 0 ELSE 1 END,
+                              CASE WHEN mf.nfc_path IS NULL THEN 0 ELSE 1 END,
+                              mf.created_at
+                      ) AS rn
+                  FROM metabase_field mf
+              )
+              UPDATE metabase_field mf
+              SET is_defective_duplicate = TRUE
+              FROM ranked_fields rf
+              WHERE mf.id = rf.id AND rf.rn > 1;
+        - sql:
+            dbms: mysql,mariadb
+            sql: >-
+              UPDATE metabase_field mf
+              LEFT JOIN (
+                  SELECT
+                      mf.id,
+                      ROW_NUMBER() OVER (
+                          PARTITION BY mf.table_id, mf.name, mf.parent_id
+                          ORDER BY
+                              CASE WHEN mf.active THEN 0 ELSE 1 END,
+                              CASE WHEN mf.nfc_path IS NULL THEN 0 ELSE 1 END,
+                              mf.created_at
+                      ) AS rn
+                  FROM metabase_field mf
+              ) x ON mf.id = x.id
+              SET mf.is_defective_duplicate = x.rn > 1;
+        - sql:
+            dbms: h2
```

Comment:
h2 query seems to be lowest common denominator for all of three on the first glance, I think it makes sense to just use it for every db then?

Reply:
sharp eyes. updated b5defa84222d5147aa53323ccc5675be5e4896e7